### PR TITLE
Enable Cross Tenant Users Resolving with Application Management

### PR DESCRIPTION
**Pull request by vinumaddumage**
### Proposed changes in this pull request
Allows cross tenant user with relevant permissions to be resolved by user ID which is required for Organization management feature

Fix Issue [wso2/product-is#14197](https://github.com/wso2/product-is/issues/14197)

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 1]
```diff
@@ -824,6 +834,99 @@ public static ServiceProvider getApplicationFromSpFileStream(SpFileStream spFile
                     "configuration file %s uploaded by tenant: %s", spFileStream.getFileName(), tenantDomain), e);
         }
     }
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
```
Comment (dewniMW) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:837]:
```suggestion
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/pom.xml [hunk 2]
```diff
@@ -222,6 +241,10 @@
                             org.wso2.carbon.consent.mgt.core.*; version="${carbon.consent.mgt.imp.pkg.version.range}",
                             org.wso2.carbon.database.utils.*;version="${org.wso2.carbon.database.utils.version.range}",
                             org.wso2.carbon.identity.central.log.mgt.*; version="${carbon.identity.package.import.version.range}",
+                            org.wso2.carbon.identity.organization.management.service; version="${org.wso2.carbon.identity.organization.management.version.range}",
+                            org.wso2.carbon.identity.organization.management.service.*; version="${org.wso2.carbon.identity.organization.management.version.range}",
+                            org.wso2.carbon.identity.organization.management.authz.service; version="${org.wso2.carbon.identity.organization.management.version.range}",
+                            org.wso2.carbon.identity.organization.management.authz.service.*; version="${org.wso2.carbon.identity.organization.management.version.range}",
```
Comment (dewniMW) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/pom.xml:247]: let's avoid using star imports

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 3]
```diff
@@ -824,6 +834,99 @@ public static ServiceProvider getApplicationFromSpFileStream(SpFileStream spFile
...
+                    Optional<org.wso2.carbon.user.core.common.User> resolvedUser =
+                            ApplicationManagementServiceComponentHolder.getInstance()
+                                    .getOrganizationUserResidentResolverService()
+                                    .resolveUserFromResidentOrganization(username, userId, accessedOrganizationId);
+                    if (resolvedUser.isPresent()) {
+                        user = new User(resolvedUser.get());
+                    }
+                }
+            }
+
```
Comment (dewniMW) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:868]: Unnecessary new line

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 4]
```diff
@@ -824,6 +834,99 @@ public static ServiceProvider getApplicationFromSpFileStream(SpFileStream spFile
                     "configuration file %s uploaded by tenant: %s", spFileStream.getFileName(), tenantDomain), e);
         }
     }
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+    public static Optional<User> getUser (String tenantDomain, String username)
```
Comment (dewniMW) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:839]:
```suggestion
    public static Optional<User> getUser(String tenantDomain, String username)
```

Comment (dewniMW) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:839]: check all applicable places

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 5]
```diff
@@ -824,6 +834,99 @@ public static ServiceProvider getApplicationFromSpFileStream(SpFileStream spFile
...
+            throws IdentityApplicationManagementException {
```
Comment (dewniMW) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:840]: add a new line after the method signature. check all applicable places

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 6]
```diff
@@ -824,6 +834,99 @@ public static ServiceProvider getApplicationFromSpFileStream(SpFileStream spFile
...
+            } else if (userId != null && userStoreManager.isExistingUserWithID(userId)) {
+                user = new User(userStoreManager.getUser(userId, null));
+            }
+        } catch (UserStoreException e) {
+            throw new IdentityApplicationManagementException("Error finding user in tenant.", e);
+        }
+        return user;
+    }
+
+    public static String getUserTenantDomain (String tenantDomain, String username)
```
Comment (dewniMW) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:893]: add method level comments to all newly introduced public methods

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java (hunk 1 above)
Comment (madurangasiriwardena) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:837]: Why this line?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 7]
```diff
@@ -824,6 +834,99 @@ public static ServiceProvider getApplicationFromSpFileStream(SpFileStream spFile
...
+            throws IdentityApplicationManagementException {
+        Optional<User> user = getUser(tenantDomain, username);
+        if (user.isPresent()) {
+            return user.get().getTenantDomain();
+        } else {
+            throw new IdentityApplicationManagementException("Error resolving user.");
```
Comment (dewniMW) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:899]: can use `orElseThrow`

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 8]
```diff
@@ -824,6 +834,99 @@ public static ServiceProvider getApplicationFromSpFileStream(SpFileStream spFile
...
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+    public static Optional<User> getUser (String tenantDomain, String username)
+            throws IdentityApplicationManagementException {
+        User user = null;
+        try {
+            int tenantID = IdentityTenantUtil.getTenantId(tenantDomain);
+            String userId = PrivilegedCarbonContext.getThreadLocalCarbonContext().getUserId();
+
+            if (tenantID == MultitenantConstants.SUPER_TENANT_ID) {
```
Comment (dewniMW) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:846]: why do we need this condition?

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 9]
```diff
@@ -824,6 +834,99 @@ public static ServiceProvider getApplicationFromSpFileStream(SpFileStream spFile
...
+                user = getUserFromTenant(username, userId, tenantID);
+            } else {
+                Tenant tenant = ApplicationManagementServiceComponentHolder.getInstance().getRealmService()
+                        .getTenantManager().getTenant(tenantID);
+                String accessedOrganizationId = tenant.getAssociatedOrganizationUUID();
+                if (accessedOrganizationId == null) {
+                    user = getUserFromTenant(username, userId, tenantID);
+                } else {
+                    if (username != null) {
```
Comment (dewniMW) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:855]: can you explain why priority is given to username here

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 10]
```diff
@@ -825,6 +835,138 @@ public static ServiceProvider getApplicationFromSpFileStream(SpFileStream spFile
         }
     }
 
+    /**
+     * Get Service provider name from XML configuration file
```
Comment (dewniMW) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:839]: invalid method description

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/User.java [hunk 11]
```diff
@@ -43,6 +43,16 @@ public class User implements Serializable {
     protected String userName;
     protected boolean isUsernameCaseSensitive = true;
 
+    public User() {
+
+    }
+
+    public User(org.wso2.carbon.user.core.common.User commonUser) {
+        this.setUserName(commonUser.getUsername());
```
Comment (dewniMW) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/model/User.java:51]: add a new line

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 12]
```diff
@@ -834,10 +976,12 @@ public static void startTenantFlow(String tenantDomain, String userName)
 
     public static void startTenantFlow(String tenantDomain) throws IdentityApplicationManagementException {
 
-        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);;
+        String userId = PrivilegedCarbonContext.getThreadLocalCarbonContext().getUserId();
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
         PrivilegedCarbonContext.startTenantFlow();
         PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain);
         PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantId(tenantId);
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setUserId(userId);
```
Comment (dewniMW) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:984]: why do we need to set the user id here?

Comment (vinumaddumage) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:975]: When a new tenant flow is started the userId cannot be accessed in that flow.

**Comment by dewniMW** [comment]
test failure fixed with https://github.com/wso2/product-is/pull/14309

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java [hunk 13]
```diff
@@ -825,6 +835,129 @@ public static ServiceProvider getApplicationFromSpFileStream(SpFileStream spFile
...
+                    Optional<org.wso2.carbon.user.core.common.User> resolvedUser =
+                            ApplicationManagementServiceComponentHolder.getInstance()
+                                    .getOrganizationUserResidentResolverService()
+                                    .resolveUserFromResidentOrganization(username, userId, accessedOrganizationId);
+                    if (resolvedUser.isPresent()) {
+                        user = new User(resolvedUser.get());
+                    }
+                }
+            }
+        } catch (UserStoreException | OrganizationManagementException e) {
```
[the remark this answers is not in the document]
Comment (vinumaddumage) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationMgtUtil.java:872]: Yes both these exceptions are thrown as a result of not being able to resolve the user.

**Pull request merged**